File: Algorithms/src/3347. Maximum Frequency of an Element After Performing Operations II/Code/solution.go
package main

import (
	"sort"
)

func maxFrequency(nums []int, k int, numOperations int) int {
	n := len(nums)
	if n == 0 {
		return 0
	}
	sort.Ints(nums)

	// frequency map
	freq := make(map[int]int, n)
	for _, x := range nums {
		freq[x]++
	}

	ans := 1

	// Case A: existing values as target
	for v, already := range freq {
		lowVal := v - k
		highVal := v + k
		L := lowerBound(nums, lowVal)
		R := upperBound(nums, highVal)
		totalInRange := R - L
		need := totalInRange - already
		canFix := need
		if canFix > numOperations {
			canFix = numOperations
		}
		if already+canFix > ans {
			ans = already + canFix
		}
	}

	// Case B: sliding window for 2*k
	l := 0
	for r := 0; r < n; r++ {
		for l <= r && nums[r]-nums[l] > 2*k {
			l++
		}
		w := r - l + 1
		cand := w
		if cand > numOperations {
			cand = numOperations
		}
		if cand > ans {
			ans = cand
		}
	}

	return ans
}

func lowerBound(a []int, target int) int {
	l, r := 0, len(a)
	for l < r {
		m := (l + r) / 2
		if a[m] < target {
			l = m + 1
		} else {
			r = m
		}
	}
	return l
}

func upperBound(a []int, target int) int {
	l, r := 0, len(a)
	for l < r {
		m := (l + r) / 2
		if a[m] <= target {
			l = m + 1
		} else {
			r = m
		}
	}
	return l
}

File: Algorithms/src/3347. Maximum Frequency of an Element After Performing Operations II/Code/solution.java
import java.util.*;

class Solution {
    public int maxFrequency(int[] nums, int k, int numOperations) {
        int n = nums.length;
        if (n == 0)
            return 0;
        Arrays.sort(nums);

        // frequency map
        Map<Long, Integer> freq = new HashMap<>();
        for (int x : nums)
            freq.put((long) x, freq.getOrDefault((long) x, 0) + 1);

        int ans = 1;

        // Case A: existing values as target
        for (Map.Entry<Long, Integer> entry : freq.entrySet()) {
            long v = entry.getKey();
            int already = entry.getValue();

            long lowVal = v - k;
            long highVal = v + k;
            int L = lowerBound(nums, lowVal);
            int R = upperBound(nums, highVal);
            int totalInRange = R - L;
            int need = totalInRange - already;
            int canFix = Math.min(need, numOperations);
            ans = Math.max(ans, already + canFix);
        }

        // Case B: sliding window for 2*k range
        int l = 0;
        for (int r = 0; r < n; ++r) {
            while (l <= r && (long) nums[r] - nums[l] > 2L * k)
                l++;
            int w = r - l + 1;
            ans = Math.max(ans, Math.min(w, numOperations));
        }

        return ans;
    }

    private int lowerBound(int[] arr, long target) {
        int l = 0, r = arr.length;
        while (l < r) {
            int mid = (l + r) >>> 1;
            if ((long) arr[mid] < target)
                l = mid + 1;
            else
                r = mid;
        }
        return l;
    }

    private int upperBound(int[] arr, long target) {
        int l = 0, r = arr.length;
        while (l < r) {
            int mid = (l + r) >>> 1;
            if ((long) arr[mid] <= target)
                l = mid + 1;
            else
                r = mid;
        }
        return l;
    }
}
